File: app/channels/conversation_channel.rb
class ConversationChannel < ApplicationCable::Channel
  def subscribed
    conversation = Conversation.find(params[:conversation_id])
    stream_for conversation
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def receive(data)
    conversation = Conversation.find(data["conversation_id"])

    message = conversation.messages.create!(
      content: data["content"],
      user: current_user
    )
  end
end

File: app/javascript/channels/conversation_channel.js
import { createConsumer } from "@rails/actioncable"

function updateMessageTimes() {
  document.querySelectorAll('.message-time').forEach(timeElement => {
    const timestamp = timeElement.dataset.timestamp
    if (timestamp) {
      const date = new Date(timestamp)
      timeElement.textContent = date.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' })
    }
  })
}

document.addEventListener('turbo:load', () => {
  const element = document.getElementById('messages')
  if (element) {
    const conversationId = element.dataset.conversationId
    createConsumer().subscriptions.create({ 
      channel: "ConversationChannel",
      conversation_id: conversationId
    }, {
      connected() {
        console.log("Connected to conversation channel")
        updateMessageTimes()
      },

      disconnected() {
        // Called when the subscription has been terminated by the server
      },

      received(data) {
        const messagesContainer = document.getElementById('messages')
        if (messagesContainer) {
          messagesContainer.innerHTML = data.html
          messagesContainer.scrollTop = messagesContainer.scrollHeight
          updateMessageTimes()
        }
      },

      speak(content) {
        this.perform('receive', { 
          conversation_id: conversationId, 
          content: content 
        })
      }
    })
  }
})
